### ISSUE TITLE
almost weekly useful materials - 2025/3/5 -

### ISSUE THREAD

**Issue by GENZITSU** [GENZITSU/UsefulMaterials#160]

**Comment by GENZITSU** [comment]
## [生成AI・AIAgentのPoCの泥沼化を避けるには？[受託開発会社向け]](https://note.com/japanmarketing/n/n89796f597f5a)


- 2025年3月現在の生成AIの出力は人のサポートなしでは完全な精度にしづらいため、社内用途が良い
- 精度を目標にすると、泥沼化しやすい。また、精度を出すための客観指標の準備も大変。
- 生成AIを用いた場合の作業時間の短縮を指標にすることで、客観指標の用意の手間を省ける。
    - 発生頻度が低い業務だと一回一回の精度が求められるし、作業時間の短縮効果も低い
    - 日頃からそこそこの頻度で発生するような業務を選ぶべき
- PoCを受託で受けず、稼働時間で受ける
    - 顧客はドメイン理解があるので、PoCをやるごとに改善アイデアがどんどん出てくる
    - 改善アイデアは確からしいのでPoCのやめ時がわからず泥沼化する
    - 精度向上施策は段々と割に合わなくなってくるため
- PoCは初期レポートに留めて、精度向上のチューニングは稼働時間ベースでもらう。
    - チューニングも受託で受けると赤字になる。
- チューニング回数を増やすためのGUI生成AIツール
    - GUI生成AIツールであれば、納品後内部でチューニングも可能
- PoCの納品物にコードや顧客が触れるシステムは含めない
    - あくまでそれらは本番開発に
- 顧客が求めているのがPoCなのかを確認する
    - 触れるシステムも欲しいのであれば、それはMVPでPoCの相場観から逸脱する。
    - PoC: 担当者がMVP開発に進むべきか判断する材料の提示
    - MVP: 担当者+数人が使えるシステムを作って、社内展開するかの判断
    - 本開発: 社内展開用にシステムを作る

### 重要なこと
- 評価指標を精度から作業時間にする
    - 作業時間に持っていくための、初期的な精度検証をどうすべきか？担当者の定性評価とか？
    - 作業時間ベースで評価する際にシステムはどこまで作り込むべきか？モック...?
- 精度向上はPoCとは別にし、稼働時間ベースの請求とする
- PoC, MVP, 本開発を分ける

**Comment by GENZITSU** [comment]
## [timmのModelEmaについて（ISIC2024コンペ振り返り①）](https://zenn.dev/morim34/articles/bfa2465defee06)

モデルの学習中にExponetial moving averageを用いることで学習の安定化を図る手法の紹介

![Image](https://github.com/user-attachments/assets/f06ac0ba-228b-4e53-ae10-cc3d498d46a5)

```python
# from https://zenn.dev/morim34/articles/bfa2465defee06

import timm
from timm.utils import ModelEmaV2

# timmから任意のモデルをロード
model = timm.create_model('resnet50', pretrained=True)

# ModelEmaV2の初期化
model_ema = ModelEmaV2(model, decay=0.9999)

# トレーニングループ内での更新
for batch in dataloader:
    optimizer.zero_grad()
    outputs = model(inputs)
    loss = loss_fn(outputs, targets)
    loss.backward()
    optimizer.step()

    # モデルのEMAを更新
    model_ema.update(model)
    
# EMAモデルの重みを保存
torch.save(model_ema.module.state_dict(), 'ema_model.pth')  
```


![Image](https://github.com/user-attachments/assets/993c60aa-e0c1-4042-a386-22366c13bb58)

ただし学習は遅くなる。

### コメント

- バッチごとにロスが激しく上下する時などには使えそう。
- learning_rateを下げておくよりかは学習が進みやすいということかな？
- transformers系だと実装がないのはなぜだろう。(pytorchのtraining loop書けば良いっちゃ良いが)

**Comment by GENZITSU** [comment]
# [Kaggle役立ちアイテム紹介（入門編）](https://speakerdeck.com/k951286/kaggleyi-li-tiaitemushao-jie-ru-men-bian?slide=13)

勉強になったスライドをペタペタ

![Image](https://github.com/user-attachments/assets/94632089-6816-4874-ab2d-662b34a0b639)

- [hugging faceをtrainerとして活用するnotebook](https://www.kaggle.com/code/ksork6s4/uspppm-bert-for-patents-baseline-train)
- [transformerで系列データを学習するnotebook](https://www.kaggle.com/code/iafoss/rna-starter-0-186-lb)
- [1DCNNで系列データを学習するnotebook](https://www.kaggle.com/code/ahmedelfazouan/belka-1dcnn-starter-with-all-data)
- [LSTM + wandabで系列データを学習するnotebook](https://www.kaggle.com/code/yasufuminakama/ventilator-pressure-lstm-starter)

![Image](https://github.com/user-attachments/assets/172cff3f-69bb-4374-986d-9ec21ec68510)

**Comment by GENZITSU** [comment]
[相互推薦システムでのPseudo Label を活用したマッチ予測精度向上の取り組み](https://speakerdeck.com/yudai00/xiang-hu-tui-jian-sisutemudenopseudo-label-wohuo-yong-sitamatutiyu-ce-jing-du-xiang-shang-noqu-rizu-m)

企業→候補者のスカウト送信, 候補者→企業のスカウト返信が両方成立する相互推薦タスクのラベルが疎になる問題を、それぞれの予測確率モデルの値から作成したPseudo Labelで緩和する手法を提案。

![Image](https://github.com/user-attachments/assets/20bf4a61-2b8d-48cf-8a95-601400b76f82)

![Image](https://github.com/user-attachments/assets/7c1602dd-64dc-4876-85ee-8dfbadad8d0f)

![Image](https://github.com/user-attachments/assets/3d4e4790-6001-4a2b-9880-a4358120707f)

分析してみると、採用担当者の行動量が多いセグメントほどPseudo Labelの効きがよく、行動量の少ないセグメントは元々の正解ラベルが重要であるという挙動を見せた。

![Image](https://github.com/user-attachments/assets/c4bce7da-212c-4c16-83f8-1dd008404e5c)

![Image](https://github.com/user-attachments/assets/3309ea14-4fda-416d-97bd-b8a05ae86628)

(本当は行動量の少ないセグメントほど、pseudo が効いた方が嬉しそう。

**Comment by GENZITSU** [comment]
## [横断検索で社内情報共有を加速させる](https://developer.hatenastaff.com/entry/2018/12/24/130000)

> ElasticsearchのインデックスにはURL、タイトル、本文、閲覧可能範囲（次節で詳しく説明）

### 情報の更新方法

2種類のjobで社内の情報をクローリングする
> 最近更新があった分をクロールしてインデックス更新するジョブ
> 業務時間内1時間に1回実行、所要時間は10分程度

> 全件をクロールしてインデックス更新するジョブ
> 業務時間外に1日1回実行、所要時間は1, 2時間程度
> データソースのAPIの仕様上最近分をインデックスをするジョブでは取りこぼす更新（閲覧権限の変更など）もこちらでカバーする

毎日、全件クローリングをやってるのが印象的 (普通なら一回きりになりそうだが、閲覧権限の変更などがあるからか?

### 検索精度を高める工夫

> Elasticsearch向けのプラグイン[elasticsearch-sudachi](https://github.com/WorksApplications/elasticsearch-sudachi)が提供されており、導入は簡単でした。次のようなDockerfileでプラグインと辞書がインストールされたElasticsearchのイメージを作成できます。

> 今回使った感触でいうと、とりわけ表記揺れに対する頑健性が役に立っています。例えば「コーヒーの淹れ方」というテキストは「淹れ方」が「入れ方」だったり「いれかた」だったりしますが、これらは全て同じトークンの列に正規化されるため、一様に検索にヒットさせることができます。他には英語の単語をアルファベットで表記するか、カタカナで表記するかの違い（「document」「ドキュメント」、「job」「ジョブ」など）も吸収してくれて便利です。「コンテキスト」「コンテクスト」のようなカタカナにしたときの表記揺れも問題ありません。

sudachiは緩めの正規化がなされていて、表記揺れに強い。(一長一短な側面あるかも)


> N-gramのサイズはbigramとtrigramをそれぞれ試し、trigramの方が検索速度で優っていたのでこちらをメインで使っています。Bigramでは、転置インデックスでの文書ヒット数が多く、その後のフレーズ検索のためのトークン位置の読み込みやスコアリングに時間がかかっているものだと思われます。

### elastic searchの設定

```json
{
  "settings": {
    "index": {
      "number_of_shards": "1",
      "number_of_replicas": "0",
      "refresh_interval": "1h",
      "analysis": {
        "analyzer": {
          "sudachi_analyzer": {
            "filter": [],
            "type": "custom",
            "tokenizer": "sudachi_tokenizer"
          },
          "ngram_analyzer": {
            "filter": [],
            "char_filter": [],
            "type": "custom",
            "tokenizer": "ngram_tokenizer"
          }
        },
        "tokenizer": {
          "sudachi_tokenizer": {
            "mode": "search",
            "resources_path": "/usr/share/elasticsearch/config/sudachi",
            "type": "sudachi_tokenizer",
            "discard_punctuation": "true"
          },
          "ngram_tokenizer": {
            "type": "nGram",
            "min_gram": "3",
            "max_gram": "3"
          }
        }
      }
    }
  },
  "mappings": {
    "_doc": {
      "dynamic": "strict",
      "properties": {
        "url": { "type": "keyword" },
        "title": {
          "type": "text",
          "analyzer": "sudachi_analyzer",
          "fields": {
            "ngram": {
              "type": "text",
              "analyzer": "ngram_analyzer"
            }
          }
        },
        "body": {
          "type": "text",
          "analyzer": "sudachi_analyzer",
          "fields": {
            "ngram": {
              "type": "text",
              "analyzer": "ngram_analyzer"
            }
          }
        },
        "permissions": { "type": "keyword" },
        "datetime": { "type": "date" }
      }
    }
  }
}
```

検索時の設定

```
{json
  "multi_match": {
    "query": "hoge",
    "type": "phrase",
    "field": ["title^2", "title.ngram", "body^2", "body.ngram"],
    "tie_breaker": 0.3
  }
}
```

> tie_breaker は title^2, title.ngram, body^2, body.ngram の中でマッチのスコアが一番が高かったもの以外のスコアも、全体のスコアで考慮させるための設定で、これを 0 より大きい値にすることで、Sudachiで同じに正規化される語でも、検索語と字面が一致するもの（n-gramでマッチするもの）が優先されます。これにより例えば「job」で検索した時に、「ジョブ」より「job」が上位に出やすくなります。

tie_breakerを用いて、sudachiのゆるい正規化のうち厳密に一致するものを優先させる。

> Elasticsearchはシングルノード、1シャード、レプリカなしで使っています。EC2インスタンスタイプはt2.mediumと小さめで、ストレージもEBSですが、それくらいの環境でも、上記設定の検索で、インデックスがファイルシステムキャッシュに乗っていれば0.5秒、そうでなくても高々2, 3秒以内で実行できていそうでした。


### コメント

6年前の記事ということが驚き。
今ならベクトル検索で済ませることも多いのだろうか。。
ベクトル検索+elastic searchのチューニングのハイブリッドとかやりうる...?

**Comment by GENZITSU** [comment]
[ML-DS-Casual-Tech-Talk 2024/02/03](https://commmune.notion.site/ML-DS-Casual-Tech-Talk-18ed641ad08281158e2fec60e3ca59a1)

Harmonic lossという新たらしいロスが紹介されている。

<img width="1207" alt="Image" src="https://github.com/user-attachments/assets/cde53cf4-7797-4357-b4ac-cb585292e261" />

<img width="1132" alt="Image" src="https://github.com/user-attachments/assets/3977f8b7-2a58-46f5-9c09-bef911c810c6" />

softmaxはwi * x の値を無限にしないといけない一方で、harmonic lossは wi - xを0にすれば良いので収束が早い。

<img width="609" alt="Image" src="https://github.com/user-attachments/assets/c331762b-a1a1-4109-bdad-d44cc9722204" />


考えうる弱点
- 入力をクラスの中心に近づけていくため、クラス間の境界が曖昧なときは微妙かも？
- ユークリッド距離で本当に良いのだろうか？

centerlossと似てるけど、あちら距離をそのままロスにしているのに対して、こちらは距離を確率に変換した上で損失を定義しているので、ちょっと違う。

https://github.com/KindXiaoming/grow-crystals

が野良実装の模様。

NNの分類手前の重みの解釈についてはこちらの記事がわかりやすい

https://qiita.com/yu4u/items/078054dfb5592cbb80cc

<img width="809" alt="Image" src="https://github.com/user-attachments/assets/e0f7c7e3-8680-43f1-a734-13f00b358f02" />

分類にしか使えなさそう。
分の埋め込みとかにはダメそう。

**Comment by GENZITSU** [comment]
[Auditing Prompt Caching in Language Model APIs](https://arxiv.org/abs/2502.07776)


ぬこぬこさん要約
> ・LLM API のプロンプトキャッシュは、データ依存のタイミング変動を引き起こし、サイドチャネルタイミング攻撃のリスクを生じさせる
> ・攻撃者は、API の応答時間の速さからキャッシュされたプロンプトを特定し、他のユーザーのプロンプトに関する情報を得ることが可能である
> ・プロンプトキャッシュの有無とキャッシュ共有レベル (ユーザーごと、組織ごと、グローバル) を特定するための統計的監査を開発・実施した
> ・OpenAI を含む 7 つの API プロバイダーで、ユーザー間でグローバルキャッシュ共有が検出され、ユーザーのプロンプトに関するプライバシー漏洩の可能性が生じた
> ・プロンプトキャッシュによるタイミング変動は、モデルアーキテクチャに関する情報漏洩も引き起こす可能性がある
> ・OpenAI の text-embedding-3-small が decoder-only Transformer であるという証拠を発見したが、これは以前は公に知られていなかった
> ・責任ある情報開示として、2024 年 10 月に監査結果を各 API プロバイダーに報告し、脆弱性に対処するための猶予期間を設けた

プロンプトキャッシュをすると、別のユーザーがどのようなpromptを入力しているかの特定をレスポンススピードを通して把握することができる。

また、prompt chachingが発生するかどうかでarchitechtureの推測ができるようになる
decoder onlyのモデルでしかprompt chachingは発生しないためである。

> プロンプトキャッシュの検出により、モデルのアーキテクチャ情報が漏れる可能性が示唆されます。デコーダーのみのTransformerでは、各トークンが過去のトークンのみを参照するため、同じ接頭辞を持つプロンプト間でキャッシュが利用され、接尾辞が異なってもヒットします。しかし、エンコーダーのみやエンコーダーデコーダー型では各トークンが全体を参照するため、このようなキャッシュは発生しません。監査では、同一接頭辞・異なる接尾辞のプロンプトでキャッシュが確認され、接尾辞の変更により出力が変わることから、キャッシュ機構が単なる出力の再利用ではないことが明らかになりました。もしtext-embedding-3-smallがTransformerベースであれば、これはデコーダーのみのTransformerであることを意味し、OpenAIがこれまで公開していなかったアーキテクチャ情報の漏洩にあたります。

**Comment by GENZITSU** [comment]
## [GraphRAG Toolkit を使って Amazon Bedrock で GraphRAG を構築する](https://acro-engineer.hatenablog.com/entry/2025/02/28/120000)

> 今回は [AWS](https://d.hatena.ne.jp/keyword/AWS) が提供する[オープンソース](https://d.hatena.ne.jp/keyword/%A5%AA%A1%BC%A5%D7%A5%F3%A5%BD%A1%BC%A5%B9)ツール [GraphRAG Toolkit](https://github.com/awslabs/graphrag-toolkit) を使って、 GraphRAG に入門しようと思います。
> GraphRAG Toolkit を使うと、 [Amazon Neptune](https://aws.amazon.com/jp/neptune/) と [Amazon OpenSearch Serverless](https://aws.amazon.com/jp/opensearch-service/features/serverless/) を使用した GraphRAG システムをローコードで実装できます。
> なお、2024年12月から Bedrock Knowledge Base で使用できるようになった GraphRAG

<img width="1225" alt="Image" src="https://github.com/user-attachments/assets/04d7709c-9425-4888-bed7-519fcbb63b40" />

> TraversalBasedRetriever にも２つの手法があり、どちらか片方、もしくは両方を組み合わせて使用することが可能です。
デフォルト設定では両方を組み合わせた結果を返却します。

> **EntityBasedSearch**
> クエリ文字列からキーワードを抽出し、マッチするエンティティをグラフから探します（Bottom up）。
> マッチするエンティティを見つけたら、そこから事実、命題、トピックと大きい概念に向かって走査することで、検索を行います。
> 未知語や固有名詞であっても、事前に整理されたナレッジグラフに従って、関連した情報を集められるのが特徴です。
> エンティティから出発するため、そのエンティティに関わる情報を幅広く取得できる傾向にあります。

> **ChunkBasedSearch**
> チャンクに対してクエリとのベクトル類似度を使いて検索を行います（Top down）。
> ヒットしたチャンクから、トピック、命題、事実と小さい概念に向かって走査することで、検索を行います。
> クエリに意味的に近いところから出発するため、より関連度の高い情報に絞って取得できる傾向にあります。


> **SemanticGuidedRetriever**
> SemanticGuidedRetriever は文言（Statement）をベクトル化したものに対して検索を行います。
> ベクトル検索で取得した文言からビームサーチとパス解析を用いて、近隣のノードから探索して関連度の高いものを取得します。
> また、リランクで質問の意図に沿った検索結果に絞る、多様性フィルターで類似した情報ばかりではなく多様性のある回答になるようにするなど、回答生成に必要な情報を検索するための仕組みが使われています。
> ベクトル検索とグラフ検索のハイブリッドを行うことで、関連度を高く保ちつつ、幅広い検索結果を返すことができますが、その分処理が重くなります。
